istioctl/pkg/precheck/precheck_test.go

```diff
@@ -0,0 +1,145 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/rest/fake"
+	cmdtesting "k8s.io/kubectl/pkg/cmd/testing"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+)
+
+type testCase struct {
+	version        string
+	revision       string
+	expectedOutput diag.Messages
+	expectedError  error
+}
+
+func Test_checkFromVersion(t *testing.T) {
+	cases := []testCase{
+		{
+			revision:       "canary",
+			version:        "1.21",
+			expectedOutput: diag.Messages{},
+			expectedError:  nil,
+		},
+		{
+			revision:       "canary",
+			version:        "XX",
+			expectedOutput: nil,
+			expectedError:  fmt.Errorf("invalid version XX, expected format like '1.0'"),
+		},
+		{
+			revision:       "canary",
+			version:        "2.0",
+			expectedOutput: nil,
+			expectedError:  fmt.Errorf("expected major version 1, got 2"),
+		},
+		{
+			revision:       "canary",
+			version:        "2.",
+			expectedOutput: nil,
+			expectedError:  fmt.Errorf("minor version is not a number: X"),
+		},
+		// TODO: add more test cases
+		// minor <= 21 and checkPilot failing
+		// minor <= 20 and checkDestinationRuleTLS failing
+		// minor <= 20 and checkExternalNameAlias failing
+		// minor <= 20 and checkVirtualServiceHostMatching failing
+		// minor <= 21 and checkPassthroughTargetPorts failing
+		// minor <= 21 and checkTracing failing
+	}
+
+	for _, c := range cases {
+		t.Run(fmt.Sprintf("revision=%s, version=%s", c.revision, c.version), func(t *testing.T) {
+			ctx := cli.NewFakeContext(&cli.NewFakeContextOption{
+				IstioNamespace: "istio-system",
+			})
+			output, err := checkFromVersion(ctx, c.revision, c.version)
+
+			assert.Equal(t, output, c.expectedOutput)
+
+			if c.expectedError != nil {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+func Test_checkTracing_ZipkinNotFound(t *testing.T) {
+	cli := kube.NewFakeClient()
+	messages := diag.Messages{}
+
+	cli.Kube().CoreV1().Services("istio-system").Create(context.Background(), nil, metav1.CreateOptions{})
+
+	err := checkTracing(cli, &messages)
+	assert.Error(t, err)
+}
+
+func Test_checkTracing_ZipkinFound(t *testing.T) {
+	cli := kube.NewFakeClient()
+	messages := diag.Messages{}
+
+	zipkinSvc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "zipkin",
+			Namespace: "istio-system",
+		},
+	}
+	cli.Kube().CoreV1().Services("istio-system").Create(context.Background(), zipkinSvc, metav1.CreateOptions{})
+
+	err := checkTracing(cli, &messages)
+	assert.NoError(t, err)
+	assert.Equal(t, 1, len(messages))
+
+	expectedOutput := msg.NewUpdateIncompatibility(ObjectToInstance(zipkinSvc),
+		"meshConfig.defaultConfig.tracer", "1.21",
+		"tracing is no longer by default enabled to send to 'zipkin.istio-system.svc'; "+
+			"follow https://istio.io/latest/docs/tasks/observability/distributed-tracing/telemetry-api/",
+		"1.21")
+
+	assert.Equal(t, expectedOutput, messages[0])
+}
+
+func init() {
```

Comment:
SHould prevent using init()

Reply:
Hey @hzxuzhonghu, thanks for the comment. Can you provide more context around this? I have taken another test in a "neighboring" package as an example and I kept the `init` from that one...

Reply:
init() is pkg scoped function. As it is run before any test case running, so we should prevent using it. To mock a pkg function, we can do that in the test case

Reply:
@hzxuzhonghu In this case the `init()` function is setting up the `cli.MakeKubeFactory` for calling `kube.NewFakeClient()` later on in the test. Do you recommend setting up `cli.MakeKubeFactory` in every test instead?

Reply:
That would not pollute the other tests, maybe not a big deal for this case

Reply:
Good to know. Thanks.